### PR TITLE
Fix translation utility class

### PR DESCRIPTION
**Pull request by ChrisTerBeke**
* Change how translation file are loaded (to increase performance, only the active locale file is loaded).
* Fixes in the I18n class.

We should add more tests for i18n, but let's do that in a separate PR.

### CODE REVIEW

File: src/components/response_error.tsx [hunk 1]
```diff
@@ -8,13 +8,6 @@ import Tile from './tile';
 // utils
 import { I18n } from '../utils/i18n';
 
-// a mapping of error codes to translated error messages which may include
-// the error parameters as replacement keys
-export type ErrorMessageTemplates = { [key: string]: string };
```
Comment (DanielSchiavini) [inline src/components/response_error.tsx]: Why not change this to `{ [key: string]: () => string },` here? We can share this for both error messages and fields, e.g. 
`export type TranslationTemplates = { [key: string]: () => string }`

File: src/components/response_error.tsx [hunk 2]
```diff
@@ -23,17 +16,28 @@ export type FieldTranslations = { [key: string]: string };
...
+    fieldTranslations: { [key: string]: () => string },
 ): string => {
     const params = { ...error };
+
+    // check if we need to translate any fields
     if (params.meta) {
         Object.keys(error.meta).forEach((key) => {
-            params[key] = (key === 'field_name' && fieldTranslations[params[key]]) || error.meta[key];
+            if (key === 'field_name' && fieldTranslations[params[key]]) {
+                // each field translation returns a function that returns a string
```
Comment (DanielSchiavini) [inline src/components/response_error.tsx]: `s/returns a function/is a function`

File: src/components/response_error.tsx (hunk 1 above)
Comment (ChrisTerBeke) [reply src/components/response_error.tsx]: It's not used that often and this reduces a 'dependency' on an import from react-web-components from the generated files, which source code lives in commons (and thus can become outdated quickly).

File: src/components/__tests__/response_error.test.tsx [hunk 3]
```diff
@@ -6,14 +6,14 @@ import { shallow } from 'enzyme';
...
 
-const ERROR_CODE_TO_I18NC_TEMPLATE: ErrorMessageTemplates = {
-    fieldError: I18n.translate('error message', '%{field_name} was invalid: %{title}.'),
+const ERROR_CODE_TO_I18NC_TEMPLATE = {
+    fieldError: () => I18n.translate('error message', '%{field_name} was invalid: %{title}.'),
 };
 
-const FIELD_NAME_TO_I18NC: FieldTranslations = {
-    username: I18n.translate('form field username', 'Username'),
+const FIELD_NAME_TO_I18NC = {
```
Comment (DanielSchiavini) [inline src/components/__tests__/response_error.test.tsx]: Now we have no type checking here and in commons

File: src/components/response_error.tsx (hunk 1 above)
Comment (DanielSchiavini) [reply src/components/response_error.tsx]: I do prefer to have it though otherwise we don't know when stuff is broken

File: src/utils/i18n.tsx [hunk 4]
```diff
@@ -1,39 +1,34 @@
...
-        // load the translation files
-        translations.forEach(async translation => (
-            this._loadTranslation(translation.name, translation.source)
-        ));
+     * Initialize the Gettext context.
+     * @param translations The translation keys with files to load.
+     */
+    public static async initialize(translations: { [s: string]: string; }): Promise<void> {
+        // register supported languages
+        Object.keys(translations).forEach((language) => {
```
Comment (DanielSchiavini) [inline src/utils/i18n.tsx]: Can't we use `Object.keys(translations).forEach(I18n._supportedLanguages.push)`?

File: src/utils/i18n.tsx [hunk 5]
```diff
@@ -1,39 +1,34 @@
...
+            I18n._supportedLanguages.push(language);
+        });
 
         // set the default locale based on the browser settings
-        I18n._gt.setLocale(I18n.getLocale());
+        const locale = I18n.getLocale();
+        if (I18n.isLanguageSupported(locale)) {
```
Comment (DanielSchiavini) [inline src/utils/i18n.tsx]: Shouldn't we have a default language when not supported? Also can't we have multiple locale in the browser? I have for example en-US, nl-NL, pt-BR, en-UK, nl-BE, pt-PT configured in this order

File: src/utils/i18n.tsx [hunk 6]
```diff
@@ -58,8 +53,8 @@ export class I18n {
     * Check if the given language is supported.
     * @param language The language to check, for example 'en'.
     */
-    public static isLanguageSupported(language: Languages): boolean {
-        return I18n._supportedLanguages.indexOf(language) > -1;
+    public static isLanguageSupported(language: string): boolean {
+        return I18n._supportedLanguages.indexOf(language) > -1 && language !== undefined;
```
Comment (DanielSchiavini) [inline src/utils/i18n.tsx]: I18n._supportedLanguages.indexOf(undefined) == -1 right, why do we need `language !== undefined`?
Note comparing to undefined doesn't work always, you need to use [typeof language == 'undefined'](https://stackoverflow.com/questions/2778901/how-to-compare-variables-to-undefined-if-i-don-t-know-whether-they-exist)

File: src/components/__tests__/response_error.test.tsx (hunk 3 above)
Comment (ChrisTerBeke) [reply src/components/__tests__/response_error.test.tsx]: There's type checking on the function itself though, this is only the test input. If it's wrong, your editor will indicate that the type doesn't match the expected type.

File: src/utils/i18n.tsx (hunk 4 above)
Comment (ChrisTerBeke) [reply src/utils/i18n.tsx]: That would be cleaner indeed.

Comment (ChrisTerBeke) [reply src/utils/i18n.tsx]: nvm, the function signatures of forEach and push don't match, so typescript doesn't allow it.

File: src/utils/i18n.tsx (hunk 5 above)
Comment (ChrisTerBeke) [reply src/utils/i18n.tsx]: The default language is english, this is already handled by the gettext library itself when registering translations. It support multi-locale fine, if there's multiple it just gets the top one from navigator.languages.

File: src/utils/i18n.tsx (hunk 6 above)
Comment (ChrisTerBeke) [reply src/utils/i18n.tsx]: You're right, that slipped through when i was testing.

**Comment by ChrisTerBeke** [comment]
@ImreBognarUltimaker can you confirm?

File: src/utils/i18n.tsx (hunk 6 above)
Comment (sedwards2009) [reply src/utils/i18n.tsx]: Comparing to `undefined` is fine here. TypeScript should prevent you from trying to access a variable which hasn't been defined, which is what Stack Overflow was talking about.

Comment (DanielSchiavini) [reply src/utils/i18n.tsx]: `undefined` is not a keyword in JS, if any library does `undefined=1` this suddenly breaks. It shouldn't happen though..

Comment (sedwards2009) [reply src/utils/i18n.tsx]: The global `undefined` variable has been read only since the Firefox 4 and ES5 era. You can't overwrite it.

Comment (DanielSchiavini) [reply src/utils/i18n.tsx]: Didn't know that, thanks!

**Pull request merged**